src/data/equiv/basic.lean

```diff
@@ -212,12 +212,50 @@ protected def ulift {α : Type u} : ulift α ≃ α :=
 protected def plift : plift α ≃ α :=
 ⟨plift.down, plift.up, plift.up_down, plift.down_up⟩
 
-@[congr] def arrow_congr {α₁ β₁ α₂ β₂ : Sort*} : α₁ ≃ α₂ → β₁ ≃ β₂ → (α₁ → β₁) ≃ (α₂ → β₂)
-| ⟨f₁, g₁, l₁, r₁⟩ ⟨f₂, g₂, l₂, r₂⟩ :=
-  ⟨λ (h : α₁ → β₁) (a : α₂), f₂ (h (g₁ a)),
-   λ (h : α₂ → β₂) (a : α₁), g₂ (h (f₁ a)),
-   λ h, by funext a; dsimp; rw [l₁, l₂],
-   λ h, by funext a; dsimp; rw [r₁, r₂]⟩
+@[congr] def arrow_congr {α₁ β₁ α₂ β₂ : Sort*} (e₁ : α₁ ≃ α₂) (e₂ : β₁ ≃ β₂) : (α₁ → β₁) ≃ (α₂ → β₂) :=
+{ to_fun := λ f, e₂.to_fun ∘ f ∘ e₁.inv_fun,
+  inv_fun := λ f, e₂.inv_fun ∘ f ∘ e₁.to_fun,
+  left_inv := λ f, funext $ λ x, by dsimp; rw [e₂.left_inv, e₁.left_inv],
+  right_inv := λ f, funext $ λ x, by dsimp; rw [e₂.right_inv, e₁.right_inv] }
+
+lemma arrow_congr_apply {α₁ β₁ α₂ β₂ : Sort*} (e₁ : α₁ ≃ α₂) (e₂ : β₁ ≃ β₂) (f : α₁ → β₁) :
```

Comment:
If you also add an `(x : α₁)` to the list of assumptions, then you can apply the composition to that `x`. In that we you can make use of coercions to functions. This has the benefit that the right hand side fits in better with the rest of the library. In particular, you could make this lemma a `@[simp]`-lemma, and it would compose well with other simp-lemmas.
This remark also applies to the other `_apply` lemma downstairs.

Reply:
This was the only `_apply` lemma without an actual application. I added `@[simp]` to some other lemmas. I can't understand why it doesn't work with `arrow_congr_comp_apply`. Could you explain this, please?

Reply:
I have no clue... I guess if you want an answer you should ask on Zulip :wink:

Reply:
@jcommelin Should be ready now. Could you review it, please?